ortools/sat/timetable_horizontal_edgefinding.cc

```diff
@@ -0,0 +1,290 @@
+  // state.
+  for (int i = 0; i < num_tasks_; ++i) {
+    if (helper_->EndMax(i) <= window_end) {
+      task_types_[i] = TaskType::kFull;
+      continue;
+    }
+    // If the task is external but has a compulsory part that starts before
+    // window_end, then we can process it partially.
+    if (helper_->StartMax(i) < window_end &&
+        helper_->StartMax(i) < helper_->EndMin(i)) {
+      task_types_[i] = TaskType::kFixedPart;
+      continue;
+    }
+    // Otherwise, simply mark the task to be ignored during sweep.
+    task_types_[i] = TaskType::kIgnore;
+  }
+
+  int next_event = 0;
+  IntegerValue time = profile_events_[0].time;
+
+  // Estimation of the earliest time at which all the processed tasks can be
+  // scheduled.
+  IntegerValue new_window_end = kMinIntegerValue;
+
+  // Overload represents the accumulated quantity of energy that could not be
+  // consumed before time.
+  IntegerValue overload = 0;
+
+  // Total demand required at time if all processed tasks were starting at
+  // their start min.
+  IntegerValue demand_req = 0;
+
+  // Total demand required at time if all processed tasks that could overlap
+  // time were. This is used to avoid placing overload in places were no task
+  // would actually be.
+  IntegerValue demand_max = 0;
+
+  while (time < window_end) {
+    // Aggregate the changes of all events happening at time. How to process
+    // an event depends on its type ("full" vs "fixed-part").
+    IntegerValue delta_max = 0;
+    IntegerValue delta_req = 0;
+    while (profile_events_[next_event].time == time) {
+      const ProfileEvent event = profile_events_[next_event];
+      switch (task_types_[event.task_id]) {
+        case TaskType::kIgnore:
+          break;  // drop the event
+        case TaskType::kFull:
+          switch (event.event_type) {
+            case ProfileEventType::kStartMin:
+              delta_max += event.height;
+              delta_req += event.height;
+              break;
+            case ProfileEventType::kEndMin:
+              delta_req -= event.height;
+              break;
+            case ProfileEventType::kEndMax:
+              delta_max -= event.height;
+              break;
+            default:
+              break;
+          }
+          break;
+        case TaskType::kFixedPart:
+          switch (event.event_type) {
+            case ProfileEventType::kStartMax:
+              delta_max += event.height;
+              delta_req += event.height;
+              break;
+            case ProfileEventType::kEndMin:
+              delta_req -= event.height;
+              delta_max -= event.height;
+              break;
+            default:
+              break;
+          }
+          break;
+      }
+      next_event++;
+    }
+
+    // Should always be safe thanks to the sentinel.
+    DCHECK_LT(next_event, profile_events_.size());
+
+    IntegerValue next_time = profile_events_[next_event].time;
+    const IntegerValue length = next_time - time;
```

Comment:
nit: please move this definition closer to where it is used.

Reply:
Given that `length` is only used once, I've inlined it directly in the overload adjustment. IMO this does not hurt the code's readability as the computation is small and pretty clear.